pkg/common/namespace_selection.go

```diff
@@ -103,3 +116,177 @@ func GetAllNamespaces(client kubernetes.Interface, labelSelector metav1.LabelSel
+	hasUpdate  bool
+	err        error
+}
+
+func (r *NamespaceSelectorReconciler) Reconcile(ctx context.Context, _ ctrl.Request) (ctrl.Result, error) {
+	log := logf.Log.WithValues("Reconciler", "NamespaceSelector")
+	oldSelections := make(map[string]NamespaceSelection)
+
+	r.lock.RLock()
+
+	for name, selection := range r.selections {
+		oldSelections[name] = selection
+	}
+
+	r.lock.RUnlock()
+
+	namespaces := corev1.NamespaceList{}
+
+	// This List will be from the cache
+	if err := r.Client.List(ctx, &namespaces); err != nil {
+		log.Error(err, "Unable to list namespaces from the cache")
+
+		return ctrl.Result{}, err
+	}
+
+	for name, oldSelection := range oldSelections {
+		newNamespaces, err := filter(namespaces, oldSelection.target)
+		if err != nil {
+			log.Error(err, "Unable to filter namespaces for policy", "name", name)
+
+			r.Update(name, NamespaceSelection{
+				target:     oldSelection.target,
+				namespaces: newNamespaces,
+				hasUpdate:  oldSelection.err != nil, // it has an update if the error state changed
+				err:        err,
+			})
+		}
+
+		if !reflect.DeepEqual(newNamespaces, oldSelection.namespaces) {
+			log.V(2).Info("Updating selection from Reconcile", "policy", name, "selection", newNamespaces)
+
+			r.Update(name, NamespaceSelection{
+				target:     oldSelection.target,
+				namespaces: newNamespaces,
+				hasUpdate:  true,
+				err:        nil,
+			})
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NamespaceSelectorReconciler) Get(name string, t policyv1.Target) ([]string, error) {
+	log := logf.Log.WithValues("Reconciler", "NamespaceSelector")
+
+	r.lock.Lock()
+
+	// If found, and target has not been changed
+	if selection, found := r.selections[name]; found && selection.target.String() == t.String() {
+		selection.hasUpdate = false
+		r.selections[name] = selection
+
+		r.lock.Unlock()
+
+		return selection.namespaces, selection.err
+	}
+
+	// unlock for now, the list filtering could take a non-trivial amount of time
+	r.lock.Unlock()
+
+	// New, or the target has changed.
+	namespaces := corev1.NamespaceList{}
+
+	// This List will be from the cache
+	if err := r.Client.List(context.TODO(), &namespaces); err != nil {
+		log.Error(err, "Unable to list namespaces from the cache")
+
+		return nil, err
+	}
+
+	selected, err := filter(namespaces, t)
+
+	log.V(2).Info("Updating selection from Reconcile", "policy", name, "selection", selected)
+
+	r.Update(name, NamespaceSelection{
+		target:     t,
+		namespaces: selected,
+		hasUpdate:  false,
+		err:        err,
+	})
+
+	return selected, err
+}
+
+func (r *NamespaceSelectorReconciler) Update(name string, sel NamespaceSelection) {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.selections[name] = sel
+}
+
+func (r *NamespaceSelectorReconciler) HasUpdate(name string) bool {
+	r.lock.RLock()
+	defer r.lock.RUnlock()
+
+	if sel, found := r.selections[name]; found {
+		return sel.hasUpdate
+	}
+
+	return true // no selections populated, call Get to populate it
```

Comment:
I think this should return false since wouldn't this make `shouldEvaluatePolicy` always return true for those without namespace selectors?

Reply:
Let me test it. I think it might still get populated on the first `Get`.

Reply:
Nevermind, I think you're right

Reply:
I addressed this in `shouldEvaluatePolicy` - when the selector wouldn't be used, it doesn't even check if the SelectorReconciler thinks here is an update (avoiding a read lock).

Reply:
I think the `evaluationInterval` CRD doc might need to change to indicate that an updated namespaceSelector result will cause the policy to be reevaluated.

Reply:
Also, `HasUpdate` is only called in `shouldEvaluatePolicy`. I think it'd be cleaner to just return `false`.

Reply:
Well, it's certainly cleaner, since if the map is missing the key, the empty value for the struct is returned, which then has `false` in the field this is looking for. So now it's a one-liner 😆 

I've updated the CRD field description.